internal/openmetrics-exporter/ports_collector.go

```diff
@@ -0,0 +1,90 @@
+package collectors
+
+import (
+	client "purestorage/fa-openmetrics-exporter/internal/rest-client"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type PortsCollector struct {
+	PortiSCSIInfoDesc   *prometheus.Desc
+	PortNVMeTCPInfoDesc *prometheus.Desc
+	PortNVMeFCInfoDesc  *prometheus.Desc
+	PortFCInfoDesc      *prometheus.Desc
+	Client              *client.FAClient
+}
+
+func (c *PortsCollector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(c, ch)
+}
+
+func (c *PortsCollector) Collect(ch chan<- prometheus.Metric) {
+	hwl := c.Client.GetPorts()
+	if len(hwl.Items) == 0 {
+		return
+	}
+	for _, h := range hwl.Items {
+		if h.Iqn != "" {
+			ch <- prometheus.MustNewConstMetric(
+				c.PortiSCSIInfoDesc,
+				prometheus.GaugeValue,
+				1,
+				h.Name, h.Iqn, h.Portal,
+			)
+		}
+		if h.Nqn != "" && h.Portal != "" {
+			ch <- prometheus.MustNewConstMetric(
+				c.PortNVMeTCPInfoDesc,
+				prometheus.GaugeValue,
+				1,
+				h.Name, h.Nqn, h.Portal,
+			)
+		}
+		if h.Nqn != "" && h.Wwn != "" {
+			ch <- prometheus.MustNewConstMetric(
+				c.PortNVMeFCInfoDesc,
+				prometheus.GaugeValue,
+				1,
+				h.Name, h.Nqn, h.Wwn,
+			)
+		}
+		if h.Wwn != "" {
+			ch <- prometheus.MustNewConstMetric(
+				c.PortFCInfoDesc,
+				prometheus.GaugeValue,
+				1,
+				h.Name, h.Wwn,
+			)
+		}
+	}
+}
+
+func NewPortsCollector(fa *client.FAClient) *PortsCollector {
+	return &PortsCollector{
+		PortiSCSIInfoDesc: prometheus.NewDesc(
+			"purefa_ports_iscsi_info",
+			"FlashArray port iscsi info",
+			[]string{"name", "iqn", "portal"},
+			prometheus.Labels{},
+		),
+		PortNVMeTCPInfoDesc: prometheus.NewDesc(
+			"purefa_ports_nvmetcp_info",
+			"FlashArray ports NVMe/TCP info",
```

Comment:
A port with this configuration could also be NVMe-RoCE . The only way to tell is from the service setting of the network port using `get_network_interfaces` and checking the `services` value, eg
```
>>> list(carray.get_network_interfaces(names=["ct0.eth4"]).items)[0].services
['nvme-tcp']
>>> list(carray.get_network_interfaces(names=["ct0.eth20"]).items)[0].services
['nvme-roce']
 ```